### PR TITLE
Classify method group assignments as methods

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/57184

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs [hunk 1]
```diff
@@ -321,6 +322,29 @@ private static bool IsInVarContext(NameSyntax name)
             return false;
         }
 
+        private static bool TryClassifyMethodGroupAssignment(
+            NameSyntax name,
+            SymbolInfo symbolInfo,
+            ArrayBuilder<ClassifiedSpan> result)
+        {
+            // If the syntax matches "var x = m" or "x = m", and there is a candidate symbol
```
Comment (JoeRobich) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:330]: Can we add a test for the `"x = m"` scenario?

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs [hunk 2]
```diff
@@ -321,6 +322,29 @@ private static bool IsInVarContext(NameSyntax name)
...
+            // for m that is a method then classify it as such.
+            if (name is IdentifierNameSyntax identifierName &&
+                symbolInfo.Symbol == null &&
+                symbolInfo.CandidateSymbols.Length == 1 &&
+                symbolInfo.CandidateReason == CandidateReason.OverloadResolutionFailure &&
+                symbolInfo.CandidateSymbols[0] is IMethodSymbol methodSymbol)
+            {
+                if (identifierName.IsRightSideOfAnyAssignExpression() || identifierName.IsVariableDeclaratorValue())
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:338]: Why is this being limited to only assignments? Because it looks like we have a bug here where I'm passing a method as an argument too?

image

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs [hunk 3]
```diff
@@ -321,6 +322,29 @@ private static bool IsInVarContext(NameSyntax name)
...
+            NameSyntax name,
+            SymbolInfo symbolInfo,
+            ArrayBuilder<ClassifiedSpan> result)
+        {
+            // If the syntax matches "var x = m" or "x = m", and there is a candidate symbol
+            // for m that is a method then classify it as such.
+            if (name is IdentifierNameSyntax identifierName &&
+                symbolInfo.Symbol == null &&
+                symbolInfo.CandidateSymbols.Length == 1 &&
+                symbolInfo.CandidateReason == CandidateReason.OverloadResolutionFailure &&
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:335]: Should we just count overload resolution as another case that sends us into the ambiguous path here?

https://github.com/dotnet/roslyn/blob/5256b889b47f5e120ad691d588c8c8b24351fe16/src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs#L72-L73

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs (hunk 2 above)
Comment (davidwengier) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:338]: It is limited only to the bugs I knew about :)

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs (hunk 3 above)
Comment (davidwengier) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:335]: Thanks for the suggestion, I will give it a try and see what breaks.

There is another spot where I thought a fix could go, where it allows overload resolution to still work for constructors, and I don't know why that wouldn't be the case for any other method, but I always assume that existing behaviour is there for a reason.

This PR is essentially a "is this fix reasonable?" question, because I didn't get an answer on Teams :)

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs (hunk 2 above)
Comment (JoeRobich) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:338]: Also, makes me wonder if method-like (Delegate, Action, Func) identifiers should be classified as method names. Totally separate conversation from this tho.

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:338]: yeah, i agree with jason.  i would even simplify this further.  if we have a SymbolInfo with an error, and a single candidate, just extract that and classify whatever we have with that symbol.

Comment (davidwengier) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:338]: > if we have a SymbolInfo with an error, and a single candidate, just extract that and classify whatever we have with that symbol.

This is a little too broad I think, and breaks a bunch of tests in not good ways. [Jason's suggestion](https://github.com/dotnet/roslyn/pull/57410/files#r739463624) only breaks 2 tests, and in ways that seem reasonable to me.

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs]: I'm curious what it breaks. Can you clarify?

Comment (davidwengier) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs]: For example, the two attributes [here](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs#L1978-L1984) are classified as classes with the change, which goes against what the comments in the test say.

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs]: IMO.  THat seems pointless.  THe code is in error, and the compiler said: welp... the best thing i can come up with is that they bound to these two symbols (illegally).   Things like f12 on the code will then take you to the besft-effort classes.

In those cases, i see no reason why it's a problem for classification to go: yeah... your code is wrong... but it looks like you did bind to these classes, so i'll classify this as a class.

--

In other words... i view the entire endeavor to *not* classify here very suspect and somewhat of a fools errand :)  It's *extra* work to stop doing a thing that is itself not bad at all :)

**Review by CyrusNajmabadi — APPROVED** [review]
Yeah. All the test changes seem totally reasonable to me.

**Comment by davidwengier** [comment]
Thank you for the assistance :)
